src/sokoban.hpp

```diff
@@ -54,10 +54,26 @@ class Sokoban {
     bool solved();
     std::vector<std::string> board();
     bool move(Direction direction);
+    bool move(unsigned int y, unsigned int x);
     bool undo();  
     bool redo();
     
     void print_board();  
 };
 
+struct PairHash {
+    template <class T1, class T2>
```

Comment:
A couple thoughts on this:

1. The `PairHash`/`PairEqual` classses seem a little bit loose and out of place here. These should probably be in its own class/file, say `SokobanCell`, which might be overkill, or a private member of `Sokoban`, since these are pretty much specific to internal Sokoban needs rather than general-purpose.
2. The templating seems unnecessary and a bit misleading. The templating makes it seem generic, like it'll work with any two types, but it actually only really works on unsigned ints, or things that make sense to bitshift like this. So I'd remove the templating and just use a header/implementation like:

```cpp
    size_t operator() (const std::pair<unsigned int, unsigned int> &p) const {
        return (p.first << 16) | p.second;
    }
```

...but I'm not super familiar with pair hash/equal overriding, so if you're doing this generically for a reason, I'd be curious to hear the rationale behind it.

Reply:
I thought about it a bit more, and I wonder if we could just have a manual conversion func (just pseudocoding here, haven't run this):

```cpp
std::unordered_map<long, std::pair<unsigned int, unsigned int>> visited;
```

Then just call this func whenever we need to hash it:

```cpp
long Sokoban::cell_key(std::pair<unsigned int, unsigned int> &cell) {
    return (cell.first << 32) | cell.second;
}
```

so:

```cpp
visited[cell_key(some_cell)] = some_other_cell;
```

Lookups would also use `cell_key(some_cell)`.

Dunno.

Another thought is that if we're shifting 16 bits, the long seems pointless, since ints are 32 bits, so shouldn't we shift 32 if we actually want to fill up the long space? It probably doesn't matter because no sokoban board will be that large, but just noticing this as a little strange.